plugin/discovery.go

```diff
@@ -0,0 +1,197 @@
+		options.providerIds = providerIds
+	}
+}
+
+// WithPluginType filters available plugins based on the plugin type
+// implemented.
+func WithPluginType(pluginType string) FindOption {
+	return func(options *findOptions) {
+		options.pluginType = pluginType
+	}
+}
+
+// FindPlugins searches for plugins in the specified directory, optionally applying filters.
+//
+// The function expects plugin manifests in the format "c2p-$PLUGIN-ID-manifest.json".
+//
+// Available filters:
+//   - `WithProviderIds`: Filters by a list of provider IDs.
+//   - `WithPluginType`: Filters by plugin type.
+//
+// If no filters are applied, all discovered plugins are returned.
+func FindPlugins(pluginDir string, opts ...FindOption) (ManifestSet, error) {
+	config := &findOptions{}
+	for _, opt := range opts {
+		opt(config)
+	}
+
+	matchingPlugins, err := findAllPluginMatches(pluginDir)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(matchingPlugins) == 0 {
+		return nil, fmt.Errorf("%w in %s", ErrPluginsNotFound, pluginDir)
+	}
+
+	metaDataSet := make(ManifestSet)
+	var errs []error
+
+	fmt.Println(matchingPlugins)
+
+	// Filter plugins by provider IDs if provided
+	if len(config.providerIds) != 0 {
+		filteredIds := make(map[string]string)
+		for _, providerId := range config.providerIds {
+			if _, ok := matchingPlugins[providerId]; !ok {
+				errs = append(errs, &NotFoundError{providerId})
+			}
+			filteredIds[providerId] = matchingPlugins[providerId]
+		}
+		matchingPlugins = filteredIds
+
+		// Return early if there are errors to avoid unnecessary processing
+		if len(errs) > 0 {
+			return nil, errors.Join(errs...)
+		}
+	}
+
+	// Process remaining plugins, filtering by plugin type if necessary
+	for id, manifestName := range matchingPlugins {
+		manifestPath := filepath.Join(pluginDir, manifestName)
+		manifest, err := readManifestFile(id, manifestPath)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if config.pluginType != "" && !manifestMatchesType(manifest, config.pluginType) {
+			continue
+		}
+
+		// sanitize the executable path in the manifest
+		cleanPath, err := sanitizeAndResolvePath(pluginDir, manifest.ExecutablePath)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		manifest.ExecutablePath = cleanPath
+		metaDataSet[id] = manifest
+	}
+
+	if len(errs) > 0 {
+		return nil, errors.Join(errs...)
+	}
+
+	if len(metaDataSet) == 0 {
+		return nil, fmt.Errorf("%w in %s with matching criteria", ErrPluginsNotFound, pluginDir)
+	}
+
+	return metaDataSet, nil
+}
+
+// findAllPluginsMatches locates the manifests in the plugin directory that match
+// the prefix name scheme and return the plugin ID and file name.
+func findAllPluginMatches(pluginDir string) (map[string]string, error) {
+	items, err := os.ReadDir(pluginDir)
+	if err != nil {
+		return nil, err
+	}
+
+	matchingPlugins := make(map[string]string)
+	for _, item := range items {
+		name := item.Name()
+		if !strings.HasPrefix(name, manifestPrefix) {
```

Comment:
@jpower432 I don't see any issues, but just wanted to double check this is fine being case sensitive?

Reply:
My thinking is that it would be straightforward if all plugins follow the same exact naming scheme to reduce the change of naming conflicts because of differentiated case. I'm curious if you see some specific trade offs here that might make it worth normalizing the file names upfront. Thoughts?

Reply:
Agreed, if we can enforce the naming scheme that would be ideal.  This was just calling out that the matching here is case sensitive, so plugin authors/installers will need to ensure all lowercase names.  I don't see any problems though.

Reply:
Yes. That is a good point. Probably something that we should highlight very clearly in documentation.